guides/4)integrating_other_frameworks/Gradio_and_Wandb_Integration.md

```diff
@@ -0,0 +1,382 @@
+        config.preserve_color = preserve_color
+        wandb.log(
+        {"Style reference": [wandb.Image(transforms.ToPILImage()(target_im))]},
+        step=0)
+        wandb.log({"Gradio panel": wandb.Html('''
+    <link rel="stylesheet" href="https://gradio.s3-us-west-2.amazonaws.com/2.6.2/static/bundle.css">
+    <div id="JoJoGAN-demo"></div>
+    <script src="https://gradio.s3-us-west-2.amazonaws.com/2.6.2/static/bundle.js"></script>
+    <script>
+    launchGradioFromSpaces("akhaliq/JoJoGAN", "#JoJoGAN-demo")
+    </script>
+    <style>
+    /* work around a weird bug */
+    .gradio_app .gradio_bg[theme=huggingface] .gradio_interface .input_dropdown .dropdown:hover .dropdown_menu {
+        display: block;
+    }
+    </style>
+    ''', inject=False)})
+
+
+    lpips_fn = lpips.LPIPS(net='vgg').to(device)
+
+
+    # reset generator
+    del generator
+    generator = deepcopy(original_generator)
+
+
+    g_optim = optim.Adam(generator.parameters(), lr=2e-3, betas=(0, 0.99))
+
+
+    # Which layers to swap for generating a family of plausible real images -> fake image
+    if preserve_color:
+        id_swap = [7,9,11,15,16,17]
+    else:
+        id_swap = list(range(7, generator.n_latent))
+
+
+    for idx in tqdm(range(num_iter)):
+        if preserve_color:
+            random_alpha = 0
+        else:
+            random_alpha = np.random.uniform(alpha, 1)
+        mean_w = generator.get_latent(torch.randn([latents.size(0), latent_dim]).to(device)).unsqueeze(1).repeat(1, generator.n_latent, 1)
+        in_latent = latents.clone()
+        in_latent[:, id_swap] = alpha*latents[:, id_swap] + (1-alpha)*mean_w[:, id_swap]
+
+
+        img = generator(in_latent, input_is_latent=True)
+        loss = lpips_fn(F.interpolate(img, size=(256,256), mode='area'), F.interpolate(targets, size=(256,256), mode='area')).mean()
+        
+        if use_wandb:
+            wandb.log({"loss": loss}, step=idx)
+            if idx % log_interval == 0:
+                generator.eval()
+                my_sample = generator(my_w, input_is_latent=True)
+                generator.train()
+                my_sample = transforms.ToPILImage()(utils.make_grid(my_sample, normalize=True, range=(-1, 1)))
+                wandb.log(
+                {"Current stylization": [wandb.Image(my_sample)]},
+                step=idx)
+            table_data = [
+                    wandb.Image(transforms.ToPILImage()(target_im)),
+                    wandb.Image(img),
+                    wandb.Image(my_sample),
+                ]
+                samples.append(table_data)
+
+
+        g_optim.zero_grad()
+        loss.backward()
+        g_optim.step()
+
+
+    out_table = wandb.Table(data=samples, columns=column_names)
+    wandb.log({"Current Samples": out_table})
+    ```
+
+
+
+    Using Web components, using the gradio-app tags allows anyone can embed Gradio demos on HF spaces directly into their blogs, websites, documentation, etc.:
```

Comment:
I'm a little confused. How come we are now talking about embedding your Jojogan Space? This wouldn't work for anyone who is following along the Guide and trying to embed their own demo.

Shouldn't we be talking about creating an `Interface` or `Blocks`, launching them, and then using the `.integrate()` function?

Reply:
thanks, added integrate and launching with interface or blocks section and moved the embedding section to the end for those that might want to do that outside of wandb